### PR TITLE
Deterministic IV for VPC

### PR DESCRIPTION
**Pull request by Taffer**
### Issues:
Resolves CryptoAlg-1205

### Description of changes: 
Adds a deterministic IV generator for use with AES-GCM; this is specifically for FIPS validation of services that require this sort of IV... even though it's not something we can FIPS validate, it needs to exist inside the FIPS boundary so the IV isn't coming from an insecure source.

### Call-outs:
n/a

### Testing:
Built FIPS/non-FIPS on macOS and ran the unit tests.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: crypto/cipher_extra/aead_test.cc [hunk 1]
```diff
@@ -1217,3 +1217,23 @@ TEST(AEADTest, WycheproofXChaCha20Poly1305) {
...
+  EXPECT_FALSE(EVP_AEAD_aes_256_gcm_det_iv_gen(0, 0, nullptr, 0));
+
+  uint8_t out[EVP_AEAD_AES_256_GCM_DET_IV_LEN] = {0};
+  EXPECT_FALSE(EVP_AEAD_aes_256_gcm_det_iv_gen(0, 0, out, 0));
+
+  uint32_t ip_address = UINT32_C(0xcdfbf267);  // amazon.com when I checked.
+  uint64_t fake_time = UINT64_C(0x1122334455667788);
+  uint8_t expected[EVP_AEAD_AES_256_GCM_DET_IV_LEN] = {
+    // TODO: This is the little-endian representation of those values; do we
+    // support any big-endian platforms?
```
Comment (torben-hansen) [inline crypto/cipher_extra/aead_test.cc:1232]: https://github.com/awslabs/aws-lc/blob/main/include/openssl/base.h#L130

File: crypto/fipsmodule/cipher/aead.c [hunk 2]
```diff
@@ -285,3 +285,21 @@ int EVP_AEAD_CTX_tag_len(const EVP_AEAD_CTX *ctx, size_t *out_tag_len,
   *out_tag_len = extra_in_len + ctx->tag_len;
   return 1;
 }
+
+// EVP_aead_aes_256_gcm_det_iv_gen computes a deterministic IV compliant with
+// NIST SP 800-38D, built from an IPv4 address and the number of nanoseconds
+// since boot, writing it to |out_iv|. It returns one on success or zero for
+// error. |out_iv_len| must be exactly 12 bytes (96 bits).
+int EVP_AEAD_aes_256_gcm_det_iv_gen(const uint32_t ip_address,
+    const uint64_t time_since_boot, uint8_t *out_iv,
```
Comment (torben-hansen) [inline crypto/fipsmodule/cipher/aead.c:294]: Should the fips module generate the time i.e. not be an input?

File: include/openssl/aead.h [hunk 3]
```diff
@@ -463,6 +467,12 @@ OPENSSL_EXPORT int EVP_AEAD_CTX_tag_len(const EVP_AEAD_CTX *ctx,
                                         const size_t in_len,
                                         const size_t extra_in_len);
 
+// EVP_aead_aes_256_gcm_det_iv_gen computes a deterministic IV compliant with
+// NIST SP 800-38D, built from an IPv4 address and the number of nanoseconds
+// since boot, writing it to |out_iv|. It returns one on success or zero for
+// error. |out_iv_len| must be exactly 12 bytes (96 bits).
+OPENSSL_EXPORT int EVP_AEAD_aes_256_gcm_det_iv_gen(const uint32_t ip_address,
```
[the remark this answers is not in the document]
Comment (Taffer) [reply include/openssl/aead.h:474]: Good point, will do.

File: include/openssl/aead.h [hunk 4]
```diff
@@ -463,6 +467,12 @@ OPENSSL_EXPORT int EVP_AEAD_CTX_tag_len(const EVP_AEAD_CTX *ctx,
...
+    const uint64_t time_since_boot, uint8_t *out_iv, const size_t out_iv_len);
```
[the remark this answers is not in the document]
Comment (Taffer) [reply include/openssl/aead.h:475]: Nope, that's just the way I'm used to doing it.

File: include/openssl/aead.h [hunk 5]
```diff
@@ -463,6 +467,12 @@ OPENSSL_EXPORT int EVP_AEAD_CTX_tag_len(const EVP_AEAD_CTX *ctx,
                                         const size_t in_len,
                                         const size_t extra_in_len);
 
+// EVP_aead_aes_256_gcm_det_iv_gen computes a deterministic IV compliant with
+// NIST SP 800-38D, built from an IPv4 address and the number of nanoseconds
```
[the remark this answers is not in the document]
Comment (Taffer) [reply include/openssl/aead.h:471]: A `uint32_t` is a canonical IP address representation, isn't it? Although they're probably in network byte order... I'm assuming the consumer of this API has the IP handy.

It's possible I'll need to massage this API to meet their needs, once I hear back from that team.

File: include/openssl/aead.h [hunk 6]
```diff
@@ -244,6 +244,10 @@ struct evp_aead_ctx_st {
 // be used.
 #define EVP_AEAD_DEFAULT_TAG_LENGTH 0
 
+// EVP_AEAD_AES_256_GCM_IV_LEN is the number of bytes required for 96 bits
+// of IV to be used in AES-GCM.
+#define EVP_AEAD_AES_256_GCM_IV_LEN 12
```
Comment (samuel40791765) [inline include/openssl/aead.h:249]: the `AES_GCM_NONCE_LENGTH` in [`crypto/fipsmodule/cipher/e_aes.c`](https://github.com/awslabs/aws-lc/blob/08b279d654fdb393c33de529e2b6244f9d6c8fdd/crypto/fipsmodule/cipher/e_aes.c) seems to imply the same value. Is it possible to reuse this?

File: crypto/fipsmodule/cipher/aead.c [hunk 7]
```diff
@@ -285,3 +285,20 @@ int EVP_AEAD_CTX_tag_len(const EVP_AEAD_CTX *ctx, size_t *out_tag_len,
   *out_tag_len = extra_in_len + ctx->tag_len;
   return 1;
 }
+
+// EVP_AEAD_iv_from_ipv4_nanosecs computes a deterministic IV compliant with
+// NIST SP 800-38D, built from an IPv4 address and the number of nanoseconds
+// since boot, writing it to |out_iv|. It returns one on success or zero for
+// error.
+int EVP_AEAD_iv_from_ipv4_nanosecs(const uint32_t ipv4_address,
+    const uint64_t nanosecs, uint8_t out_iv[12]) {
```
Comment (samuel40791765) [inline crypto/fipsmodule/cipher/aead.c:294]: might want to also update `out_iv[12]` to take the defined variable

File: include/openssl/aead.h (hunk 6 above)
Comment (Taffer) [reply include/openssl/aead.h:249]: Good call! I'd been looking for a useful `#define`, but only in the `.h` files.

**Review by samuel40791765 — DISMISSED** [review]

File: crypto/fipsmodule/cipher/aead.c [hunk 8]
```diff
@@ -285,3 +285,20 @@ int EVP_AEAD_CTX_tag_len(const EVP_AEAD_CTX *ctx, size_t *out_tag_len,
   *out_tag_len = extra_in_len + ctx->tag_len;
   return 1;
 }
+
+// EVP_AEAD_iv_from_ipv4_nanosecs computes a deterministic IV compliant with
+// NIST SP 800-38D, built from an IPv4 address and the number of nanoseconds
+// since boot, writing it to |out_iv|. It returns one on success or zero for
+// error.
+int EVP_AEAD_iv_from_ipv4_nanosecs(const uint32_t ipv4_address,
```
Comment (samuel40791765) [inline crypto/fipsmodule/cipher/aead.c:293]: slight naming nit, but 
```suggestion
int EVP_AEAD_get_iv_from_ipv4_nanosecs(const uint32_t ipv4_address,
```
would be nice, and aligns with the naming of the `EVP_AEAD_CTX_get_iv` function above (Line 262).

**Review by samuel40791765 — APPROVED** [review]

**Review by sdkellyAWS — DISMISSED** [review]

File: include/openssl/aes.h [hunk 9]
```diff
@@ -67,6 +67,8 @@ extern "C" {
 
 #define AES_BLOCK_SIZE 16
 
+#define AES_GCM_NONCE_LENGTH 12
```
Comment (torben-hansen) [inline include/openssl/aes.h:70]:
```suggestion
#define FIPS_AES_GCM_NONCE_LENGTH 12
```
and define it in `aead.h`. Or `#define EVP_AEAD_AES_GCM_NONCE_LENGTH 12`

The nonce of AES-GCM doesn't have to be 12 bytes long.

Comment (Taffer) [reply include/openssl/aes.h:70]: This nonce has to be 12 bytes...

**Comment by torben-hansen** [comment]
Pending confirmation of signature of `EVP_AEAD_get_iv_from_ipv4_nanosecs()` is OK.

**Comment by Taffer** [comment]
> Pending confirmation of signature of `EVP_AEAD_get_iv_from_ipv4_nanosecs()` is OK.

Based on reading the CryptoBR tickets linked in CryptoAlg-1205, I think this is good.

File: include/openssl/aead.h [hunk 10]
```diff
@@ -15,6 +15,7 @@
 #ifndef OPENSSL_HEADER_AEAD_H
 #define OPENSSL_HEADER_AEAD_H
 
+#include <openssl/aes.h>
```
Comment (torben-hansen) [inline include/openssl/aead.h:18]: Redundant now?

Comment (Taffer) [reply include/openssl/aead.h:18]: The best thing about this code review is how there are more comments than changed/added lines. ;-)

Right you are!

**Review by torben-hansen — APPROVED** [review]

**Pull request merged**